api.js

```diff
@@ -306,6 +306,42 @@ module.exports = function attachAPI(app, {io, db}) {
     }
   ])
 
+  app.post('/api/pin-message', [
+    ...middleware.loadVarFromBody('messageID'),
+    ...middleware.loadVarFromBody('sessionID'),
+    ...middleware.getSessionUserFromID('sessionID', 'sessionUser'),
+    ...middleware.requireBeAdmin('sessionUser'),
+    ...middleware.getMessageFromID('messageID', 'message'),
+    (req, res, next) => {
+      const v = req[middleware.vars]
+      v.channelID = v.message.channelID
+      next()
+    },
+    ...middleware.getChannelFromID('channelID', 'channel'),
+
+    async (request, response) => {
+      const { messageID, channel } = request[middleware.vars]
+
+      if (channel.pinnedMessageIDs.includes(messageID)) {
+        response.status(500).end(JSON.stringify({
+          error: 'this message is already pinned'
+        }))
+
+        return
+      }
+
+      await db.channels.update({_id: channel._id}, {
+        $push: {
+          pinnedMessageIDs: messageID
```

Comment:
May be worth checking to see if the message ID is actually a valid message ID and is in that channel!

Reply:
To be fair only administrators can pin messages, but I still think it's worth checking beforehand so we don't get any edge cases in the future if we forget to check there.

Reply:
Actually, both of those things are already done.

A message is always pinned in the channel it's contained in. Notice how there's no `channelID` parameter? That's because the channel is automatically fetched from the passed message:

```js

+    (req, res, next) => {
+      const v = req[middleware.vars]
+      v.channelID = v.message.channelID
+      next()
+    },
+    ...middleware.getChannelFromID('channelID', 'channel'),
+ 
```

The message is always valid, because this line (gracefully, i.e. 404) fails if it isn't:

```js
 +    ...middleware.getMessageFromID('messageID', 'message'), 
```

Reply:
Oman I didn't even notice that -- _awesome_!